test3: Add md.wait() and second run_random()

dlsloan opened the pull request:
test3 runs on disks which are not mounted in md with --assume-clean so
are not guarenteed to have a clean array before testing is finished.
This means that fast-path io may not be tested. Fix this by adding a
wait and second run_random to tests.

lsgunth reviewed (commented):
Actually, this was always the default, so basic tests don't bother with syncing. Though it got confusing when the python arguments changed.

test3 was supposed to disable assume-clean and run IO during sync, and presumably some of that IO would continue after sync was complete. I'd prefer that to continue to be the case. Though the naming should probably be fixed up.

dlsloan commented:
Ah, can we rename it to --no-assume-clean and add an option+run of ./test3 --verify --assume-clean to test all then since there's no guarentee that data will be written before the array is fully built.

lsgunth commented:
Yes, that sounds good to me.

lsgunth commented:
You'll need this fix:

```diff
diff --git a/md.py b/md.py
index 3036263f2176..4dc4907adbd2 100644
--- a/md.py
+++ b/md.py
@@ -39,7 +39,7 @@ class _EnvironmentArgMixin:
             self._env_found = env
 
         envval = os.environ.get(env, action.default)
-        if action_type == "store_true":
+        if action_type in ("store_true", "store_false"):
             envval = self.to_bool(envval)
         nargs = kwargs.get("nargs", None)
         if ((nargs in ("+", "*") or isinstance(nargs, int)) and
```

Then you can do something like this:

```python3
    parser.add_argument("-D", "--no-def", dest="def", action="store_false")
    parser.add_argument("-d", "--def", action="store_true")
```

The order of the two arguments will set which is default. If --no-def is first, the default is True, otherwise false.

lsgunth commented:
Actually, on second thought, instead of adding a another run to test_all, just change test3 to wait and run more tests...

That ensures there will be IO run when through the fast path, with less time added to our testing.

Something like this:

```python
diff --git a/test3 b/test3
index 4118e304d63b..0fc0c34c6ee5 100755
--- a/test3
+++ b/test3
@@ -125,6 +125,10 @@ class TestRunner:
             if quick:
                 return
 
+    def wait_for_clean(self):
+        self.log(f"Waiting for clean")
+        self._md.wait()
+
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
     group = parser.add_mutually_exclusive_group()
@@ -146,6 +150,8 @@ if __name__ == "__main__":
             cnt = 6
 
         runner.run_random(cnt)
+        runner.wait_for_clean()
+        runner.run_random(cnt)
 
         if args.grow_test:
             runner.grow_tests(cnt, args.quick)
```

dlsloan committed:
test3: Add md.wait and second run random

test3 runs on disks which are not mounted in md with --assume-clean so
are not guarenteed to have a clean array before testing is finished.
This means that fast-path io may not be tested. Fix this by adding a
wait and second run_random to tests.

dlsloan commented:
Used your last patch verbatum and tested on savitar

lsgunth reviewed (approved):
Looks good. Also worth fixing up that assume-clean confusion if you have a chance.
